src/lm/mikolov-rnnlm-lib.cc

```diff
@@ -0,0 +1,1084 @@
+// Tomas Mikolov are licensed under the Apache License, the short form
+// os which is below.  The original code by Tomas Mikolov is licensed
+// under the BSD 3-clause license, whose text is further below.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+// Original BSD 3-clause license text:
+// Copyright (c) 2010-2012 Tomas Mikolov
+//
+// All rights reserved. Redistribution and use in source and binary forms, with
+// or without modification, are permitted provided that the following conditions
+// are met: 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following
+// disclaimer. 2. Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the
+// distribution. 3. Neither name of copyright holders nor the names of its
+// contributors may be used to endorse or promote products derived from this
+// software without specific prior written permission. THIS SOFTWARE IS PROVIDED
+// BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR
+// IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
+// EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
+// OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+// EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <assert.h> 
+#include <stdio.h> 
+#include <stdlib.h> 
+#include <string.h> 
+#include <math.h> 
+#include <time.h> 
+#include "lm/mikolov-rnnlm-lib.h"
+#include "util/table-types.h"
+
+///// fast exp() implementation
+static union {
+  double d;
+  struct {
+    int j,i;
+  } n;
+} d2i;
+#define EXP_A (1048576/M_LN2)
+#define EXP_C 60801
+#define FAST_EXP(y) (d2i.n.i = EXP_A*(y)+(1072693248-EXP_C),d2i.d)
+
+///// include blas
+#ifdef USE_BLAS
+extern "C" {
+#include  < cblas.h > 
```

Comment:
Am I right in assuming that this is not ever used and could be deleted?  Seems a bit confusing.